[Finder] add iname and notIname methods

jschaedl opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28503
| License       | MIT
| Doc PR        | -

I added the methods `$finder->files()->iname()` and `$finder->files()->notIname()` to be able to filter files case-insensitively.

jschaedl committed:
[Finder] add iname and notIname methods

xabbuh commented on src/Symfony/Component/Finder/Glob.php:
```diff
@@ -45,7 +45,7 @@ class Glob
      * @return string regex The regexp
      */
-    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#')
+    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#', $isCaseSensitive = true)
```
we can add a `bool` type hint for the new argument

jschaedl committed:
[Finder] add type hint

xabbuh commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -153,6 +155,32 @@ public function date($dates)
+     * You can use patterns (delimited with / sign), globs or simple strings.
+     *
+     *     $finder->iname('*.php')
+     *     $finder->iname('/\.php$/') // same as above
```
As far as I understand it with the current implementation the two examples above are not fully equivalent (the glob will match `foo.PHP` while the regex wouldn't), right? Does it make sense to modify any passed regex by adding the `i` modifier?

jschaedl replied:
That's right the glob `$finder->iname('*.php')` will match `*.php` and `*.PHP`. The regex should be `/\.php$/i` to be the same. Will fix it.

Adding the `i` modifier to every regex passed to `$finder->iname()` makes sense to me. Will add it.

jschaedl committed:
[Finder] add case-insensitive modifier to regex passed to iname and notIname method

debreczeniandras commented on src/Symfony/Component/Finder/Glob.php:
```diff
@@ -45,7 +45,7 @@ class Glob
      * @return string regex The regexp
      */
-    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#')
+    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#', bool $isCaseSensitive = true)
```
Wouldn't it be useful to use a `string $modifier = ''` parameter here, that would allow passing further modifiers to this class?

jschaedl replied:
I wouldn't do that. I would rather add more boolean parameters for other modifiers in case they are needed in the future.

nicolas-grekas replied:
I'd name this arg $caseSensitive

debreczeniandras commented on src/Symfony/Component/Finder/Iterator/FilenameFilterIterator.php:
```diff
@@ -42,6 +51,6 @@ public function accept()
     protected function toRegex($str)
     {
-        return $this->isRegex($str) ? $str : Glob::toRegex($str);
+        return $this->isRegex($str) ? $str.($this->isCaseSensitive ? '' : 'i') : Glob::toRegex($str, true, true, '#', $this->isCaseSensitive);
```
What if we already received the 'i' modifier from the i/name method?
Shouldn't that be checked here to avoid the duplicate modifier?

debreczeniandras commented:
Accidentally I also started working on the same issue and ran into the same problems like Travis.
(with the folder permission errors)

I created a pull request for the fix: #28582.

Commit:
[Finder] Restore folder permissions after failed assertions in FinderTest suite

add CaseInsensitiveFilenameFilterIteratorTest

jschaedl committed:
[Finder] prevent case-insensitive modifier to be applied twice in toRegex

jschaedl commented:
@stof 

Atm the pattern passed to `$finder->name()` will always win, because the `FilenameFilterIterator` of  the `names` array is wrapped in the `FilenameFilterIterator` of the `inames` array. See https://github.com/symfony/symfony/blob/aa4a9a2cf3c614db0be88c7037c33d655ee1b4da/src/Symfony/Component/Finder/Finder.php#L759

I think I wouldn't use both `name()` and `iname()` in the same finder. But their might be use cases which I don't see atm and we should combine the handling of the `names` and `inames` patterns in one `FilenameFilterIterator` rather than splitting it in two different ones? WDYT?

nicolas-grekas commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -714,7 +757,11 @@ private function searchInDirectory(string $dir): \Iterator
 
         if ($this->names || $this->notNames) {
-            $iterator = new Iterator\FilenameFilterIterator($iterator, $this->names, $this->notNames);
+            $iterator = new Iterator\FilenameFilterIterator($iterator, $this->names, $this->notNames, true);
```
this change is not needed, it should be reverted to reduce merge conflicts

nicolas-grekas commented on src/Symfony/Component/Finder/Glob.php:
```diff
@@ -111,6 +111,6 @@ public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardS
         }
 
-        return $delimiter.'^'.$regex.'$'.$delimiter;
+        return $delimiter.'^'.$regex.'$'.$delimiter.((!$isCaseSensitive) ? 'i' : '');
```
no need for the internal brackets

nicolas-grekas commented on src/Symfony/Component/Finder/Iterator/MultiplePcreFilterIterator.php:
```diff
@@ -101,6 +101,25 @@ protected function isRegex($str)
         return false;
     }
 
+    protected function hasRegexModifiers(string $str, array $modifiers)
```
I'm not sure we need this logic: let's just add "i" and not care about duplicates, that'd behave the same in the end, isn't it?

jschaedl replied:
That's right, duplicates behave the same.

nicolas-grekas commented on src/Symfony/Component/Finder/Tests/FinderTest.php:
```diff
@@ -37,7 +37,8 @@ public function testFiles()
         $this->assertSame($finder, $finder->files());
-        $this->assertIterator($this->toAbsolute(array('foo/bar.tmp',
+        $this->assertIterator($this->toAbsolute(array(
+            'foo/bar.tmp',
```
please revert change

nicolas-grekas commented on src/Symfony/Component/Finder/Tests/FinderTest.php:
```diff
@@ -49,13 +50,17 @@ public function testFiles()
         $finder->files();
-        $this->assertIterator($this->toAbsolute(array('foo/bar.tmp',
+        $this->assertIterator($this->toAbsolute(array(
+            'foo/bar.tmp',
```
please revert

nicolas-grekas commented on src/Symfony/Component/Finder/Tests/FinderTest.php:
```diff
@@ -126,7 +137,8 @@ public function testDepth()
         $this->assertSame($finder, $finder->depth('< 1'));
-        $this->assertIterator($this->toAbsolute(array('foo',
+        $this->assertIterator($this->toAbsolute(array(
+            'foo',
```
to revert

nicolas-grekas commented on src/Symfony/Component/Finder/Tests/FinderTest.php:
```diff
@@ -138,11 +150,15 @@ public function testDepth()
         $this->assertSame($finder, $finder->depth('<= 0'));
-        $this->assertIterator($this->toAbsolute(array('foo',
+        $this->assertIterator($this->toAbsolute(array(
+            'foo',
```
revert :)

jschaedl committed:
[Finder] revert some changes to reduce merge conflicts

jschaedl committed:
[Finder] rename case sensitive parameter

jschaedl committed:
[Finder] remove hasRegexModifiers method and apply i modifier directly

jschaedl commented:
@nicolas-grekas 

> Adding a new argument would be a BC break so this asks for a BC layer.

Changing the method signature to `public function name($patterns, $ignoreCase = false)` wouldn't be a BC break, right? Or did I miss something?

> And adding a boolean argument isn't very clear about what it means for the reader.

Actually I agree on that, but for the special use case of filtering with case-insensitivity, I could live with it.

nicolas-grekas commented:
> Changing the method signature to public function name($patterns, $ignoreCase = false) wouldn't be a BC break,

yes it would: it would break any classes extending Finder (see https://symfony.com/bc also)

jschaedl commented:
> yes it would: it would break any classes extending Finder (see https://symfony.com/bc also)

Ah right, thanks for pointing that out.

stof commented:
@jschaedl it could still allow mixing them, as you can call this method multiple times to register multiple name patterns (that's why the Finder has an array of names)

xabbuh commented on src/Symfony/Component/Finder/Glob.php:
```diff
@@ -42,10 +42,11 @@ class Glob
      * @return string regex The regexp
      */
-    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#')
+    public static function toRegex($glob, $strictLeadingDot = true, $strictWildcardSlash = true, $delimiter = '#', bool $caseSensitive = true)
```
we cannot add the argument here as that is a BC break for any class extending this one

xabbuh commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -153,6 +155,31 @@ public function date($dates)
+     *
+     * @see FilenameFilterIterator
+     */
+    public function iname($patterns)
```
we could add a return type hint

xabbuh commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -192,6 +219,22 @@ public function notName($patterns)
+     *
+     * @see FilenameFilterIterator
+     */
+    public function notIname($patterns)
```
here too

jschaedl committed:
[Finder] code review changes

- add return type hints to iname and notIname methods
- move caseSensitive handling completely into FilenameFilterIterator

dunglas commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -153,6 +155,31 @@ public function date($dates)
+     *
+     * You can use patterns (delimited with / sign), globs or simple strings.
+     *
+     *     $finder->iname('*.php')
```
These examples aren't very easy to understand.

Maybe something like:

```php
*     $finder->iname('*.php') // will match both *.php and *.PHP
```

?

javiereguiluz reviewed (approved):
Ready too ... but let's simplify the phpdoc of `iname()` as suggested by Kévin. Thanks!

jschaedl committed:
[Finder] Update phpdoc iname method

nicolas-grekas commented on src/Symfony/Component/Finder/Finder.php:
```diff
@@ -153,6 +155,30 @@ public function date($dates)
+     * You can use patterns (delimited with / sign), globs or simple strings.
+     *
+     *     $finder->iname('*.php') // will match both *.php and *.PHP
+     *     $finder->iname('/\.php$/i') // same as above
```
Wait, this means we can already achieve case insensitivity right? which means we don't need `iname()`!

jschaedl replied:
Yes, that's right. If you use a regex, you can achieve case insensitivity with `$finder->name(...)`
This possible solution was described in the description of #28503.

fabpot replied:
I would prefer to not add `iname()` as it makes things more complex for something we can already do with the current `name()` method.

nicolas-grekas commented:
> I would prefer to not add iname() as it makes things more complex for something we can already do with the current name() method.

Closing as explained. Thanks for working on it and making it a viable proposal!

MrBLT commented:
> This is unfortunate. Not everyone knows or are comfortable with regexs. Also, case insensitivity in file searches is more common than case-sensitive searches. I'm searching specifically because I'm not the one creating the files and can't guarantee the case. That's why this component is so fantastic versus raw PHP functions. At a minimum the documentation should be updated as I searched the documentation for "insensitive" (again, a very, very common requirement in for this use case) and didn't get any insight.

Agreed. This should be case insensitive by default, and/or there should be a function for `$finder->files()->extension('php')` which would be case insensitive. Or the Docs need updated to specify that the examples shown are not.

I'm working with hundreds of thousands of drawings and found out some of them aren't showing up in my utility because I followed Symfony Docs first suggestion that appeared to find files by extension `$finder->files()->name('*.php')`. every Symfony Docs (or code) example within Finder that references `->name('*.php')` implies that it's going to find every PHP file.  Which it doesn't, if someone created the php file as *.Php or *.PHP (which is still a very valid file-extension from what I have read).

I guess since I'm in a Windows environment I'm biased by their filesystem... So for someone in in a linux/unix environment, are you able to make a new folder and add two files to it, one named `foo.bar` and the other named `foo.Bar` and successfully have two different files? If so then it makes sense why it is the way it is!! But if not, then `foo.bar` = `foo.Bar` and they should both show up using a simple search string '*.bar'

fabpot commented:
> > > I would prefer to not add iname() as it makes things more complex for something we can already do with the current name() method.
> > 
> > 
> > Closing as explained. Thanks for working on it and making it a viable proposal!
> 
> This is unfortunate. Not everyone knows or are comfortable with regexs. Also, case insensitivity in file searches is more common than case-sensitive searches. I'm searching specifically because I'm not the one creating the files and can't guarantee the case. That's why this component is so fantastic versus raw PHP functions. At a minimum the documentation should be updated as I searched the documentation for "insensitive" (again, a very, very common requirement in for this use case) and didn't get any insight.

What about suggesting some improvements to the documentation?
